Handle arXiv errors

benureau opened the pull request:
The idea is to be as helpful and precise to the user when something fails. Also:
- handles old/new style arxiv references
- normalize arXiv citekeys
- improves error handling of doi/isbn (raise Exception if an HTTP status is detected).
- fix the isbn parse test that was incorrectly using `doi2bibtex`.
- add feedparser to the requirements.txt file.
- allows to run the tests using `pytest`.

benureau commented:
@joe-antognini: here is the new PR building on your work, if you are interested. 
@omangin: I still have some work to do. I will tell you when everything is ready.

omangin commented:
Good! Don't forget to add a `changelog` entry and to update the list of contributors.

benureau commented:
Oh, yes, I did not think about that. Btw, I requested the review, but it's not ready yet.

benureau commented:
Ok, I addressed #154, and partially #147.

@omangin: I renamed `requirements.txt` into `dev_requirements.txt` and added comments inside to avoid confusions.

benureau commented:
@omangin: ok, did some work on parsing pyparsing errors. Needs further tests.

omangin commented on tests/test_apis.py:
```diff
@@ -2,19 +2,42 @@
 
     def setUp(self):
+        if not _is_connected():
+            self.skipTest('no connection detected, skiping test')
```
Can't we use a `@unittest.skipIf(not _is_connected)` instead? This would make it more explicit. And we could also have tests in the API that don't use the network. That's a good way to do it too, so it's fine anyway if you like it more.

benureau replied:
With the mock properly implemented, not needed anymore.

benureau commented on pubs/commands/add_cmd.py:
```diff
@@ -41,22 +43,25 @@ def parser(subparsers, conf):
                     question='Bibfile not edited. Edit again ?',
                     default='y')
                 if not again:
                     ui.exit(0)
```
Hum... I don't know, I view that as a correct termination of the program there. The user, for whatever reason, decided to end the edit without committing the changes. When closing vim with "q!", the return code is 0.

omangin replied:
OK. That kind of make sense. I also remember that the _similar case_ I had in mind is actually slightly different, so it is not inconsistent to keep this, at least for now.

benureau commented on pubs/endecoder.py:
```diff
@@ -1,9 +1,16 @@
 try:
     import bibtexparser as bp
+    # don't let bibtexparser display stuff
+#    bp.bparser.logger.setLevel(level=logging.CRITICAL)
```
Nice catch. It should not be commented. Bibtexparser should not log on stdout: we should control how errors are displayed.

benureau commented on pubs/endecoder.py:
```diff
@@ -68,14 +75,16 @@ class EnDecoder(object):
     class BibDecodingError(Exception):
 
-        message = "Could not parse provided bibdata:\n---\n{}\n---"
+#        message = "Could not parse provided bibdata:\n---\n{}\n---"
```
Yes. Kept it until I was sure.

benureau commented on dev_requirements.txt:
```diff
@@ -0,0 +1,23 @@
+# if you want to setup your environment for development of the pytest code,
+# doing `pip install -r dev_requirements.txt` is the single thing you have to do.
+# Alternatively, and perhaps more conveniently, running `python setup.py test`
+# will do the same *and* run the tests.
```
agreed.

benureau commented on pubs/apis.py:
```diff
@@ -1,27 +1,210 @@
+               "'{}': {}".format(arxiv_id, e))
+        raise ReferenceNotFoundError(msg)
+
+    # print("TEXT = '{}'".format(r.text))
```
done.

benureau commented on pubs/apis.py:
```diff
@@ -1,27 +1,210 @@
+        msg = 'the arXiv server returned an error message: {}'.format(entry['summary'])
+        raise ReferenceNotFoundError(msg)
+    # import pprint
+    # pprint.pprint(entry)
```
done.

benureau commented on pubs/apis.py:
```diff
@@ -1,27 +1,210 @@
+    bibtex = bibtexparser.dumps(db)
+    return bibtex
+
+if __name__ == '__main__':
```
done.

benureau commented:
The last commit handles #147

benureau commented on pubs/apis.py:
```diff
@@ -161,12 +160,6 @@ def arxiv2bibtex(arxiv_id, try_doi=True, ui=None):
     entry = feed.entries[0]
-    if 'arxiv.org/api/errors' in entry['id']:  # server is returning an error message.
-        msg = 'the arXiv server returned an error message: {}'.format(entry['summary'])
-        raise ReferenceNotFoundError(msg)
```
@omangin: Just so you understand why I removed this. I talked with the arXiv devs. This would have generated a 400 HTTP status code, and already been handled above.

benureau commented on tests/mock_requests.py:
```diff
@@ -0,0 +1,89 @@
+
+import os
+try:
+    import cPickle as pickle
```
Wait wait. `cPickle` is not a dependency. It's the name of the the builtin C implementation of the pickle module in Python 2.7. Python 3 renamed cPickle into pickle and (only?) has the C-implementation. This is just meant to use the faster version for Python 2.7. Probably not that significant a speed boost, but I thought it would mean using more similar implementation between Python 2 and 3, but on reading the doc, Python 2.7's pickle and cPickle actually produce identical info (https://docs.python.org/2/library/pickle.html#module-cPickle).

Compatibility between Python 2 and 3 is ensured by calling `dump` with `protocol=2`, the latest protocol supported by Python 2.7.

About using json versus pickle, well, pickle generates smaller files, that are read faster. Current tests generate 43K of data, pickled. I used pickle to avoid the repository size getting large as we modify tests.

omangin replied:
>About using json versus pickle, well, pickle generates smaller files, that are read faster. 

Are you sure about that? I don't see any reason why it would be the case since the data is mostly text. Also, it is easy to compress so should not affect the repository size in a sensible way. And if we kind of sort the data before dumping, the json file should also be versioned easily so will impact less the repository size than a binary file that is modified each time.

There is also a serious security concern related to using pickle which is that it would be equivalent to potentially distributing code that is not directly readable. Even if we trust ourselves, we would have to re-generate the pickle file ourselves each time somebody contributes a commit that changes it to be able to trust its content. We could organize around that but it seems that there are many concerns here for a very speculative gain.

benureau commented on tests/mock_requests.py:
```diff
@@ -0,0 +1,89 @@
+            r.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            error_msg = str(e)
+        key = (sorted(args), sorted((k, v) for k, v in kwargs.items()))
```
Good eyes. Artifact from an earlier code. Horrible stuff.

omangin replied:
I thought about making the arguments hashable to use a mapping too… and to implement a mapping out of a sorted list…  ;)

benureau commented:
Yes, I know you would say that. Which means I need to modify the Travis matrix. Man, I have nightmare of the ten+ commits it took me to make it work last time I tried to do that.

omangin commented:
Regarding the Travis matrix, I was just thinking of running both the online and offline test for each environment (by directly modifying the test `script` option). That's a pretty trivial change. Also since the run time consists mostly in the loading of the environment, it should not impact it so much. We could introduce a tag for tests that use connections to re-run only these if we want to optimize.

benureau commented:
Yep, that what I figured.

benureau commented:
Ok, I'll add more tests tomorrow, and i think we are done.

omangin commented on tests/mock_requests.py:
```diff
@@ -81,8 +77,8 @@ def mock_requests_get(*args, **kwargs):
-        with open(os.path.join(_data_filepath), 'wb') as fd:
-            pickle.dump(_collected_responses, fd, protocol=2)
+        with open(os.path.join(_data_filepath), 'w') as fd:
+            json.dump(_collected_responses, fd)
```
We might want to use `indent=2`  in the Json dump. This falls back to deciding whether the data is part of the test *documentation* and whether it is readable. On one side it would make the file slightly bigger, on the other it would better be diffed in git. If we chose this path, we probably want to sort the responses first.
```python
_collected_responses.sort()
json.dump(_collected_responses, fd, indent=2)
```
This is pretty minor, so I leave it to you to decide.

benureau replied:
agreed.

omangin commented on .travis.yml:
```diff
@@ -36,8 +36,11 @@ matrix:
+  - export PUBS_TESTS_MODE=ONLINE
+  - python setup.py test
+  - export PUBS_TESTS_MODE=MOCK
+  - python setup.py test
```
Maybe:
```yaml
script: 
  - PUBS_TESTS_MODE=ONLINE python setup.py test
  - PUBS_TESTS_MODE=MOCK python setup.py test
```

benureau replied:
Ok, I went with:
```
script:
  - PUBS_TESTS_MODE=MOCK python setup.py test
  - PUBS_TESTS_MODE=COLLECT python setup.py test
```
to test the collect code.

benureau commented:
@omangin: I'll let you merge this one once if you don't have any problem with the last additions.

omangin commented on tests/mock_requests.py:
```diff
@@ -43,6 +45,19 @@ def raise_for_status(self):
+        if '10.1103/PhysRevD.89.084044' in text:
+            # replace with wrong DOI
+            text = text.replace('PhysRevD', 'INVALIDDOI')
+    except TypeError:
```
Actually I can't reproduce the case in which the `try` is needed. In any case using a `str('PhysRevD')` or something similar would also be an option.

benureau replied:
I don't think this has to do with unicode, but with bytes versus strings. `Response.text` in request is a property, so it can't be reassigned. To make the `ONLINE` version work, I modify the `r._content` variable, which is bytes. If you have an alternative, I am interested.

omangin replied:
OK, I see. My point was that there was maybe a simple way to avoid duplicating the body of the function in the catch block. But there is probably none and it is fine like that for testing code.
